knowledge_base/framework/complex-action.md

```diff
@@ -0,0 +1,358 @@
+            chain_id: finalComment.chain,
+            author_address: finalComment.Address.address,
+            author_chain: finalComment.Address.chain,
+          },
+        },
+        excludeAddresses: [finalComment.Address.address],
+      });
+    });
+  }
+```
+
+  `Repository`
+
+```typescript
+  // update address last active
+  address.last_active = new Date();
+  address.save();
+```
+  
+  `Schemas`
+
+```typescript
+  return [finalComment.toJSON(), allNotificationOptions];
+}
+```
+
+## Solution
+
+We can establish a straightforward "conventional framework" designed to delineate business requirements from the technical implementation details. By leveraging well-established behavioral and creational design patterns like the chain of responsibility, builder, and repository, we can boost development efficiency and code robustness. This framework should also support the automated generation of a significant portion of the code and structure necessary for implementing and documenting new features.
+
+### Components
+
+By adopting the hexagonal architecture approach, a top level application port can take on the role of configuring routes with authentication and schema validations, further delegating various other responsibilities to other dedicated ports, each handling tasks related to persistence, messaging, and more.
+
+We can use the "builder" pattern to initialize the application with various adapters tailored for specific use cases such as unit testing, local development, and production environments.
+
+Additionally, we can employ convention-over-configuration techniques to define the commands we expose through the API, including their schemas and implementation specifics.
+
+```typescript
+export const updateComment:Command<Request,Response> = {
+  auth: request => {...},
+  body: z.object(...),
+  response: z.object(...),
+  callback: (actor, body) => {
+    ...
+  }
+}
+```
+
+Cross-cutting concerns, such as the choice of API protocol (REST, gRPC) and data repository (in-memory, database dialect), are established within the bootstrapping routine.
+
+```typescript
+import { commands } from "./commands"
+
+app(Express(options));
+
+app().with(commands).build();
+app().listen()
+```
+
+The pseudocode representation of a refactored handler might appear as follows, leaving the developer solely responsible for implementing the business callback interface:
+
+```typescript
+const command = (actor, body, callback) => {
+  // These steps are guaranteed by the framework, including spec documentation
+  // 1. Authorize Command Execution - *`Auth`*
+  // 2. Message Schema Validation - *`Schemas`*
+  // 3. Route Command - *`Routing`*
+
+  try {
+    // CORE business - callback configured
+    const state, events = callback(actor, body) {
+      // 4. Load Aggregate - *`Repository`*
+      const state = await repo().loadBy(...)
+      // 5. Validate Invariants (Business Rules) - *`Business Invariants`*
+      check invariants or raise Error
```

Comment:
Are business invariants different from request invariants? If we perform all validation in step 5, we will have wasted a db call in step 4 assuming that the shape of the request is incorrect.

Reply:
Yes, request invariants are under "schema validation and auth" responsibilities. Business invariants (or aggregate invariants) on the other hand are making sure the "state" of the aggregate is valid before and after accepting the command